File: 155-min-stack/min-stack.cpp
class MinStack {
public:
    stack<int>s;
    stack<int>s1;
    MinStack() {
        
    }
    
    void push(int val) {
       s.push(val); 
       s1.push(min(val, s1.empty() ? val : s1.top()));
    }
    
    void pop() {
        s.pop();
        s1.pop();
    }
    
    int top() {
        return s.top();
    }
    
    int getMin() {
        return s1.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.mini = float('inf')
        self.stack = []

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append(val)
            self.mini = val
        elif val >= self.mini:
            self.stack.append(val)
        else:
            self.stack.append(2 * val - self.mini)
            self.mini = val

    def pop(self) -> None:
        if not self.stack:
            return
        x = self.stack.pop()
        if x < self.mini:
            self.mini = 2 * self.mini - x

    def top(self) -> int:
        if not self.stack:
            return
        x = self.stack[-1]  
        if x >= self.mini:
            return x
        else:
            return self.mini  

    def getMin(self) -> int:
        return self.mini